File: y2024/day19/day19.go
package day19

import (
	"adventofcode/y2024/utils"
	"fmt"
	"strings"
)

const SAMPLE = `r, wr, b, g, bwu, rb, gb, br

brwrr
bggr
gbbr
rrbgbr
ubwu
bwurrg
brgr
bbrgwb`

var MEMO = map[string]bool{}
var MEMO_WAYS = map[string]int{}

type Input struct {
	Puzzles []string
	Stripes []string
}

func (i *Input) ProcessInput(fpath string) {
	content := utils.Read(fpath)
	// content = SAMPLE
	input := strings.Split(content, "\n\n")
	i.Stripes = strings.Split(input[0], ", ")
	i.Puzzles = strings.Split(input[1], "\n")
}

func NewInput(daydir string) Input {
	i := Input{}
	i.ProcessInput(daydir + "/input.txt")
	return i
}

func matchStripes(stripes []string, puzzleStr string) bool {
	if _, ok := MEMO[puzzleStr]; ok {
		return MEMO[puzzleStr]
	}

	if puzzleStr == "" {
		return true
	}

	for _, word := range stripes {
		if strings.HasPrefix(puzzleStr, word) {
			matched := matchStripes(stripes, puzzleStr[len(word):])
			if matched {
				MEMO[puzzleStr] = true
				return true
			}
		}
	}
	MEMO[puzzleStr] = false
	return false
}

func countWays(stripes []string, puzzleStr string) int {
	if _, ok := MEMO_WAYS[puzzleStr]; ok {
		return MEMO_WAYS[puzzleStr]
	}
	if puzzleStr == "" {
		return 1
	}

	ways := 0
	for _, word := range stripes {
		if strings.HasPrefix(puzzleStr, word) {
			ways += countWays(stripes, puzzleStr[len(word):])
		}
	}
	MEMO_WAYS[puzzleStr] = ways
	return ways
}

func Part01(input Input) {
	numSolvables := 0
	for _, puzzle := range input.Puzzles {
		isSolvable := matchStripes(input.Stripes, puzzle)
		if isSolvable {
			numSolvables++
		}
	}
	fmt.Printf("PART 01: %d\n", numSolvables)
}

func Part02(input Input) {
	numWays := 0
	for _, puzzle := range input.Puzzles {
		ways := countWays(input.Stripes, puzzle)
		numWays += ways
	}
	fmt.Printf("PART 02: %d\n", numWays)
}

func Run(dir string) {
	utils.PartPrinter("DAY 19")
	input := NewInput(dir + "/day19")
	Part01(input)
	Part02(input)
}

File: y2024/day19/day19.py
import os
from typing import Literal


SAMPLE = """r, wr, b, g, bwu, rb, gb, br

brwrr
bggr
gbbr
rrbgbr
ubwu
bwurrg
brgr
bbrgwb"""


class Input:
    def __init__(self, filepath: str):
        self.stripes : dict[str, Literal[True]]
        self.stripe_length: set[int]
        self.longest_stripe: int
        self.puzzle: list[str]
        self.__post_init__(filepath)

    def __post_init__(self, filepath: str):
        with open(filepath, "r") as fp:
            content = fp.read()
        # content = SAMPLE
        self.stripes, self.puzzle = content.split("\n\n")
        self.scan_window = max([len(stripe) for stripe in self.stripes.split(", ")])
        self.stripe_length = set(map(len, self.stripes))
        self.stripes = {k: True for k in self.stripes.split(", ")}
        self.puzzle = self.puzzle.split("\n")


MEMO: dict[str, int] = {}


def match_stripe(stripes: dict[str, bool], puzzle_str: str) -> bool:
    if MEMO.get(puzzle_str):
        return MEMO[puzzle_str]
    if not puzzle_str:
        return True
    for word in stripes.keys():
        if puzzle_str.startswith(word):
            matched = match_stripe(stripes, puzzle_str.removeprefix(word))
            if matched:
                MEMO[puzzle_str] = matched
                return True
    MEMO[puzzle_str] = False
    return False


def part01(data: Input):
    num_solvable = 0
    for pzl in data.puzzle:
        print(f"{pzl=}", end=", ", flush=True)
        is_match = match_stripe(data.stripes, pzl)
        print(f"{is_match=}")
        if is_match:
            num_solvable += 1
    print(f"{num_solvable=}")


def part02(data: Input):
    ...


def main():
    print("""
U   U  SSS  EEEE       ggg      ooo
U   U S     E         g        o   o
U   U  SSS  EEE      g   gggg  o   o
u   U     S E         g   g g  o   o
 UUU   SSS  EEEE       ggg  g   ooo

    """)
    os._exit(0)
    data = Input("./day19/input.txt")
    part01(data)
    part02(data)


if __name__ == "__main__":
    main()

